# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
import scipy.stats

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents. 

In [2]:
# Your code here:
df=pd.read_csv('Current_Employee_Names__Salaries__and_Position_Titles.csv')


Examine the `salaries` dataset using the `head` function below.

In [3]:
# Your code here:
df


Name                 Job Titles        Department  \
0      ADRIANO,  RACQUEL ANNE                  PARAMEDIC              FIRE   
1           AFFANEH,  MAHIR A             POLICE OFFICER            POLICE   
2           AFOENYI,  PHYLLIS  TITLE V PROGRAM TRAINEE I  FAMILY & SUPPORT   
3             AGAR,  BULENT B        DEPUTY COMMISSIONER       WATER MGMNT   
4             AGEE,  LAWRENCE             POLICE OFFICER            POLICE   
...                       ...                        ...               ...   
33697      ZYLINSKA,  KLAUDIA             POLICE OFFICER            POLICE   
33698      ZYMANTAS,  LAURA C             POLICE OFFICER            POLICE   
33699       ZYMANTAS,  MARK E             POLICE OFFICER            POLICE   
33700     ZYRKOWSKI,  CARLO E             POLICE OFFICER            POLICE   
33701     ZYSKOWSKI,  DARIUSZ    CHIEF DATA BASE ANALYST              DAIS   

      Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
0                     F           Salary            NaN        72510.0   
1                     F           Salary            NaN        93354.0   
2                     P           Hourly           20.0            NaN   
3                     F           Salary            NaN       132972.0   
4                     F           Salary            NaN        90024.0   
...                 ...              ...            ...            ...   
33697                 F           Salary            NaN        72510.0   
33698                 F           Salary            NaN        76266.0   
33699                 F           Salary            NaN        90024.0   
33700                 F           Salary            NaN        93354.0   
33701                 F           Salary            NaN       126732.0   

       Hourly Rate  
0              NaN  
1              NaN  
2             13.0  
3              NaN  
4              NaN  
...            ...  
33697          NaN  
33698          NaN  
33699          NaN  
33700          NaN  
33701          NaN  

[33702 rows x 8 columns]

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [4]:
# Your code here:
df.isna().sum()


Name                     0
Job Titles               0
Department               0
Full or Part-Time        0
Salary or Hourly         0
Typical Hours        25528
Annual Salary         8174
Hourly Rate          25528
dtype: int64

Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [5]:
# Your code here:
df['Salary or Hourly'].value_counts()


Salary    25528
Hourly     8174
Name: Salary or Hourly, dtype: int64

What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [6]:
# Your code here:
df_departments_count=df.groupby(['Department']).size().reset_index(name='Number of Employees')
df_departments_count.head(40)

Department  Number of Employees
0          ADMIN HEARNG                   38
1         ANIMAL CONTRL                   75
2              AVIATION                 1856
3     BOARD OF ELECTION                  111
4       BOARD OF ETHICS                    8
5         BUDGET & MGMT                   44
6             BUILDINGS                  258
7      BUSINESS AFFAIRS                  167
8            CITY CLERK                   88
9          CITY COUNCIL                  362
10                 COPA                  123
11     CULTURAL AFFAIRS                   74
12                 DAIS                 1033
13         DISABILITIES                   28
14     FAMILY & SUPPORT                  656
15              FINANCE                  544
16                 FIRE                 4627
17               HEALTH                  467
18              HOUSING                   66
19   HOUSING & ECON DEV                  154
20      HUMAN RELATIONS                   17
21      HUMAN RESOURCES                   75
22        INSPECTOR GEN                   95
23                  LAW                  398
24    LICENSE APPL COMM                    1
25       MAYOR'S OFFICE                  104
26                 OEMC                 1737
27               POLICE                13848
28         POLICE BOARD                    2
29          PROCUREMENT                   84
30       PUBLIC LIBRARY                 1088
31  PUBLIC SAFETY ADMIN                  244
32        STREETS & SAN                 2175
33           TRANSPORTN                 1113
34            TREASURER                   28
35          WATER MGMNT                 1914

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [7]:
# Your code here:  8174 employees on hourly rate 
print('The company´s average hourly rate is:', round(np.mean(df['Hourly Rate']), 2))


The company´s average hourly rate is: 34.59


In [8]:
hour_rate_sample = df.loc[df['Hourly Rate'].notna(), 'Hourly Rate']


In [9]:
from scipy.stats import ttest_1samp

In [10]:
confidence_level = 0.95
degrees_freedom = len(hour_rate_sample)
sample_mean = np.mean(hour_rate_sample)

In [11]:
sample_standard_error = scipy.stats.sem(hour_rate_sample)
confidence_interval = scipy.stats.t.interval(confidence_level, 
                                             degrees_freedom, 
                                             sample_mean, 
                                             sample_standard_error)

In [12]:
print( 'confidence interval is ', confidence_interval, '.' )

confidence interval is  (34.2983953972632, 34.874504039977715) .


In [13]:
# Null hypothesis or H0: mean hourly rate value >= 30
# Alternative hyp or H1: mean hourly rate value < 30

In [14]:
stat, pval = ttest_1samp(hour_rate_sample, 30)

In [15]:
print('stat is  ', stat)
print('pvalue for the two-tailed test is ', pval)

print('pvalue for the one-tailed test is ', pval/2)

stat is   31.211500329216044
pvalue for the two-tailed test is  3.825475847673794e-202
pvalue for the one-tailed test is  1.912737923836897e-202


In [16]:
30 - np.mean(hour_rate_sample)

-4.586449718620457

In [17]:
scipy.stats.sem(hour_rate_sample)*stat

4.586449718620017

> Conclusion: the company´s hourly rate is 4.6 dollars above the 30 dollars/h.

### We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that 
#salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. 
#Test this one sided hypothesis using a 95% confidence interval.

#Hint: A one tailed test has a p-value that is half of the two tailed p-value. 
#If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [18]:
# Your code here:

police_dept_by_salaries = df.loc[df['Department'].str.contains('POLICE'), ['Department', 'Annual Salary']]


In [19]:
police_dept_by_salaries['Annual Salary'] = police_dept_by_salaries['Annual Salary'].dropna()


In [20]:
police_dept_by_salaries['Annual Salary'] = pd.to_numeric(police_dept_by_salaries['Annual Salary'], errors='coerce')


In [21]:
police_dept_by_salaries = police_dept_by_salaries.dropna()


In [22]:
police_dept_by_salaries['Annual Salary'].isna().sum()

0

In [23]:
police_dept_by_salaries_array = np.array(police_dept_by_salaries['Annual Salary'])


In [24]:
#### testin

In [25]:
confidence_level = 0.95
degrees_freedom = len(police_dept_by_salaries_array)
sample_mean = np.mean(police_dept_by_salaries_array)
sample_standard_error = scipy.stats.sem(police_dept_by_salaries_array)
confidence_interval = scipy.stats.t.interval(confidence_level, 
                                             degrees_freedom, 
                                             sample_mean, 
                                             sample_standard_error)

In [26]:
stat, pval = ttest_1samp(police_dept_by_salaries_array, 86000)

In [27]:
print('stat is  ', stat)
print('pvalue for the two-tailed test is ', pval)

print('pvalue for the one-tailed test is ', pval/2)

stat is   18.519537603296808
pvalue for the two-tailed test is  1.178882027843547e-75
pvalue for the one-tailed test is  5.894410139217735e-76


In [28]:
86000 - np.mean(police_dept_by_salaries_array)


-2834.256328656149

In [29]:
scipy.stats.sem(police_dept_by_salaries_array)*stat

2834.25632865615

> Conclusion: The average salary for the police force is 2834 dollars more than 86000 dollars of last year.

Using the `crosstab` function, find the department that has the most hourly workers. 

In [49]:
# Your code here:

data_crosstab = pd.crosstab(df['Department'], df['Hourly Rate'], margins = True)

data_crosstab    #  from the column All we can see the max count is at STREETS & SAN

Hourly Rate          2.65  8.25  9.74  10.0  10.5  12.0  13.0  13.4  13.45  \
Department                                                                   
ANIMAL CONTRL           0     0     0     0     0     0     0     0      0   
AVIATION                0     0     0     0     0     0     0     0      0   
BUDGET & MGMT           0     0     0     0     0     0     2     0      0   
BUSINESS AFFAIRS        0     0     0     0     0     0     2     0      0   
CITY COUNCIL            0     4     0     2     1     5     5     0      1   
CULTURAL AFFAIRS        0     0     0     0     0     0     0     0      0   
DAIS                    0     0     0     0     0     0     0     0      0   
DISABILITIES            0     0     0     0     0     0     1     0      0   
FAMILY & SUPPORT      260     0     0     0     0     0    41     0      0   
FINANCE                 0     0     0     0     0     0     0     0      0   
HEALTH                  0     0     0     0     0     0     0     0      0   
HOUSING & ECON DEV      0     0     0     0     0     0     0     0      0   
LAW                     0     0     0     0     0     0     0     0      0   
MAYOR'S OFFICE          0     0     0     0     0     0     0     0      0   
OEMC                    0     0     0     0     0     0     0     0      0   
POLICE                  0     0     2     0     0    11    11     0      0   
PROCUREMENT             0     0     0     0     0     0     0     0      0   
PUBLIC LIBRARY          0     0     0     0     0     0     0   107      0   
PUBLIC SAFETY ADMIN     0     0     0     0     0     0     0     0      0   
STREETS & SAN           0     0     0     0     0     0     0     0      0   
TRANSPORTN              0     0     0     0     0     0     0     0      0   
WATER MGMNT             0     0     0     0     0     0     0     0      0   
All                   260     4     2     2     1    16    62   107      1   

Hourly Rate          13.94  ...  53.4  53.91  54.05  54.1  55.1  57.04  58.4  \
Department                  ...                                                
ANIMAL CONTRL            0  ...     0      0      0     0     0      0     0   
AVIATION                 0  ...     0      0      0     0     0     12     0   
BUDGET & MGMT            0  ...     0      0      0     0     0      0     0   
BUSINESS AFFAIRS         0  ...     0      0      0     0     0      0     0   
CITY COUNCIL             0  ...     0      0      0     0     0      0     0   
CULTURAL AFFAIRS         0  ...     0      0      0     0     0      0     0   
DAIS                     0  ...     0      0      1    24     4     10     0   
DISABILITIES             0  ...     0      0      0     0     0      0     0   
FAMILY & SUPPORT         0  ...     0      0      0     0     0      0     0   
FINANCE                  0  ...     0      0      0     0     0      0     0   
HEALTH                   0  ...     0      0      0     0     0      0     0   
HOUSING & ECON DEV       0  ...     0      0      0     0     0      0     0   
LAW                      0  ...     0      0      0     0     0      0     0   
MAYOR'S OFFICE           0  ...     0      0      0     0     0      0     0   
OEMC                    49  ...     0      0      0     0     0      0     0   
POLICE                   0  ...     0      0      0     0     0      0     0   
PROCUREMENT              0  ...     0      0      0     0     0      0     0   
PUBLIC LIBRARY           0  ...     0      0      0     0     0      0     0   
PUBLIC SAFETY ADMIN      0  ...     9      0      0     0     0      0     0   
STREETS & SAN            0  ...     0      0      0     0     3      0     0   
TRANSPORTN               0  ...    39      0      1     0     1      0    10   
WATER MGMNT              0  ...     0      1     28     0     4     46     0   
All                     49  ...    48      1     30    24    12     68    10   

Hourly Rate          61.65  128.0   All  
Depart

> The Department that has the most hourly workers is: STREETS & SAN

The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [42]:
df_filtered = df[df['Hourly Rate'].notnull()]   ## much better code for slicing/filtering. We gotta remove the nulls in 
# order to be able to run the t-test

In [43]:
# Your code here:
streets_hourly = df_filtered.loc[df_filtered['Department'].str.contains('STREETS & SAN'), ['Department', 'Hourly Rate']]



In [44]:
streets_hourly_array = np.array(streets_hourly['Hourly Rate'])


In [45]:
confidence_level = 0.95
degrees_freedom = len(streets_hourly_array)
sample_mean = np.mean(streets_hourly_array)
sample_standard_error = scipy.stats.sem(streets_hourly_array)
confidence_interval = scipy.stats.t.interval(confidence_level, 
                                             degrees_freedom, 
                                             sample_mean, 
                                             sample_standard_error)

In [46]:
stat, pval = ttest_1samp(streets_hourly_array, 35)

In [47]:
print('stat is  ', stat)
print('pvalue for the two-tailed test is ', pval)

print('pvalue for the one-tailed test is ', pval/2)

stat is   4.000138803968834
pvalue for the two-tailed test is  6.580474443773579e-05
pvalue for the one-tailed test is  3.2902372218867895e-05


In [48]:
scipy.stats.sem(streets_hourly_array)*stat

0.5522968580715002

> Conclusion: The workers are unfortunately wrong, because their average hourly/rate is # $0.55 above average.

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [59]:
# Your code here: # using the code from exercise below, with all the hourly rate values (without nulls)
#hour_rate_sample 


In [61]:
from scipy import stats

In [57]:
hour_rate_sample_mean= np.mean(hour_rate_sample)

In [62]:
standard_error = stats.sem(hour_rate_sample)

In [71]:
confidence_interval = stats.t.interval(0.95, len(hour_rate_sample)-1, hour_rate_sample_mean, standard_error)


In [73]:
print("The 95% confidence interval for the average hourly wage is:", confidence_interval)


The 95% confidence interval for the average hourly wage is: (34.29839539204361, 34.874504045197305)


Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

In [52]:
# Your code here:  # retriving the array already created on the exercises above.
police_dept_by_salaries_array


array([ 93354.,  90024., 111444., ...,  76266.,  90024.,  93354.])

In [67]:
confidence_level = 0.95
degrees_freedom = len(police_dept_by_salaries_array)-1
sample_mean = np.mean(police_dept_by_salaries_array)
sample_standard_error = scipy.stats.sem(police_dept_by_salaries_array)
confidence_interval2 = scipy.stats.t.interval(confidence_level, 
                                             degrees_freedom, 
                                             sample_mean, 
                                             sample_standard_error)

In [74]:
print("The 95% confidence interval for the average salary in the police force is:", confidence_interval2)

The 95% confidence interval for the average salary in the police force is: (88534.27439465115, 89134.23826266115)


# Bonus Challenge - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [ ]:
# Your code here:

